sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java

```diff
@@ -0,0 +1,605 @@
+    } else {
+      byte[] bytes = value.toJavaBigDecimal().unscaledValue().toByteArray();
+      WritableColumnVector arrayData = toColumn.getChildColumn(0);
+      arrayData.putBytes(index * 16, bytes.length, bytes, 0);
+      toColumn.putArray(index, index * 16, bytes.length);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // Put repeating values
+  // --------------------------------------------------------------------------
+
+  private void putByteArrays(int count, WritableColumnVector toColumn, byte[] bytes) {
+    WritableColumnVector arrayData = toColumn.getChildColumn(0);
+    int size = bytes.length;
+    arrayData.reserve(size);
+    arrayData.putBytes(0, size, bytes, 0);
+    for (int index = 0; index < count; index++) {
+      toColumn.putArray(index, 0, size);
+    }
+  }
+
+  private static void putDecimalWritables(
+      WritableColumnVector toColumn,
+      int size,
+      int precision,
+      int scale,
+      HiveDecimalWritable decimalWritable) {
+    HiveDecimal decimal = decimalWritable.getHiveDecimal();
+    Decimal value =
+      Decimal.apply(decimal.bigDecimalValue(), decimal.precision(), decimal.scale());
+    value.changePrecision(precision, scale);
+
+    if (precision <= Decimal.MAX_INT_DIGITS()) {
+      toColumn.putInts(0, size, (int) value.toUnscaledLong());
+    } else if (precision <= Decimal.MAX_LONG_DIGITS()) {
+      toColumn.putLongs(0, size, value.toUnscaledLong());
+    } else {
+      byte[] bytes = value.toJavaBigDecimal().unscaledValue().toByteArray();
+      WritableColumnVector arrayData = toColumn.getChildColumn(0);
+      arrayData.reserve(bytes.length);
+      arrayData.putBytes(0, bytes.length, bytes, 0);
+      for (int index = 0; index < size; index++) {
+        toColumn.putArray(index, 0, bytes.length);
+      }
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // Put non-null values
+  // --------------------------------------------------------------------------
+
+  private void putNonNullBooleans(int count, LongColumnVector fromColumn, WritableColumnVector toColumn) {
+    long[] data = fromColumn.vector;
+    for (int index = 0; index < count; index++) {
+      toColumn.putBoolean(index, data[index] == 1);
+    }
+  }
+
+  private void putNonNullByteArray(int count, BytesColumnVector fromColumn, WritableColumnVector toColumn) {
+    BytesColumnVector data = fromColumn;
+    WritableColumnVector arrayData = toColumn.getChildColumn(0);
+    int totalNumBytes = IntStream.of(data.length).sum();
+    arrayData.reserve(totalNumBytes);
+    for (int index = 0, pos = 0; index < count; pos += data.length[index], index++) {
+      arrayData.putBytes(pos, data.length[index], data.vector[index], data.start[index]);
+      toColumn.putArray(index, pos, data.length[index]);
+    }
+  }
+
+  private void putNonNullBytes(int count, LongColumnVector fromColumn, WritableColumnVector toColumn) {
```

Comment:
shall we revert these? Since they do not help with performance, I'd like to inline these small functions.

Reply:
Yep. I'll revert them.

Reply:
It almost seemed that they hurt the performance. The MR and Hive tests were up and down randomly as expected, but Vectorized tests were down in almost every benchmark.